src/gpl/src/mbff.cpp

```diff
@@ -0,0 +1,1024 @@
+#include <lemon/maps.h>
+using namespace std;
+using namespace lemon;
+using namespace operations_research;
+using namespace sat;
+
+
+using Graph = ListDigraph;
+using Node = ListDigraph::Node;
+using Edge = ListDigraph::Arc;
+
+auto seed = chrono::high_resolution_clock::now().time_since_epoch().count();
+std::mt19937 mt(seed);
+
+/* hard coded values for height, width, power consumption, etc. */
+float WIDTH = 2.448, HEIGHT = 1.2;
+float RATIOS[6] = {1, 0.875, 0.854, 0.854, 0.844, 0.844}, POWER[6] = {1, 0.875, 0.854, 0.854, 0.844, 0.844};
+int BITCNT[6] = {1, 4, 8, 16, 32, 64};
+namespace gpl {
+	// begin or-tools usage 
+
+		float MBFF::RunLP(vector<Flop> flops, vector<Tray> &trays, vector<pair<int, int> > clusters, int sz) {
+			int num_flops = (int)flops.size(), num_trays = (int)trays.size();
+
+			unique_ptr<MPSolver> solver(MPSolver::CreateSolver("GLOP"));
+			const double inf = solver->infinity();
+
+			vector<const MPVariable*> tray_x(num_trays), tray_y(num_trays);
+			for (int i = 0; i < num_trays; i++) {
+				tray_x[i] = solver->MakeNumVar(0, inf, "");
+				tray_y[i] = solver->MakeNumVar(0, inf, "");
+			}
+
+			vector<const MPVariable*> d_x(num_flops), d_y(num_flops);
+
+
+			for (int i = 0; i < num_flops; i++) {	
+
+				d_x[i] = solver->MakeNumVar(0, inf, "");
+				d_y[i] = solver->MakeNumVar(0, inf, "");
+
+
+				int tray_idx = clusters[i].first, slot_idx = clusters[i].second;
+
+				Point flop = flops[i].pt;
+				Point tray = trays[tray_idx].pt;
+				Point slot = trays[tray_idx].slots[slot_idx]; 
+
+
+				float shift_x = slot.x - tray.x, shift_y = slot.y - tray.y;
+
+				MPConstraint* const c1 = solver->MakeRowConstraint(shift_x - flop.x, inf, "");
+				c1->SetCoefficient(d_x[i], 1);
+				c1->SetCoefficient(tray_x[tray_idx], -1);
+
+
+				MPConstraint* const c2 = solver->MakeRowConstraint(flop.x - shift_x, inf, "");
+				c2->SetCoefficient(d_x[i], 1);
+				c2->SetCoefficient(tray_x[tray_idx], 1);
+
+				MPConstraint* const c3 = solver->MakeRowConstraint(shift_y - flop.y, inf, "");
+				c3->SetCoefficient(d_y[i], 1);
+				c3->SetCoefficient(tray_y[tray_idx], -1);
+
+				MPConstraint* const c4 = solver->MakeRowConstraint(flop.y - shift_y, inf, "");
+				c4->SetCoefficient(d_y[i], 1);
+				c4->SetCoefficient(tray_y[tray_idx], 1);
+			}
+
+			MPObjective* const objective = solver->MutableObjective();
+			for (int i = 0; i < num_flops; i++) {
+				objective->SetCoefficient(d_x[i], 1);
+				objective->SetCoefficient(d_y[i], 1);
+			}
+			objective->SetMinimization();
+
+
+
+
+			float tot_d = 0;
+			for (int i = 0; i < num_trays; i++) {
+				Tray new_tray;
+				new_tray.pt.x = tray_x[i]->solution_value();
+				new_tray.pt.y = tray_y[i]->solution_value();
+			    tot_d += MBFF::GetDist(trays[i].pt, new_tray.pt);
+		        trays[i] = new_tray;
+			 }
+
+			return tot_d;
+		}
+
+
+		/* 
+			NOTE: CP-SAT constraints only work with INTEGERS
+			so, all coefficients (shift_x and shift_y) are multiplied by 100
```

Comment:
Can we use DBU in the internal solver. Those units are integer by default.

Reply:
To my understand, shift_x and shift_y will always be floats because they involve slot locations (which will be floats because of the ratio multiplier).